contracts/exchangeIssuance/FlashMintNotional.sol

```diff
@@ -0,0 +1,901 @@
+        }
+
+        (address[] memory componentsBought, uint256[] memory amountsBought) =  _buyComponentsForInputToken(
+            _tradeData,
+            _swapData
+        );
+
+        _mintWrappedFCashPositions(
+            _tradeData,
+            componentsBought,
+            amountsBought
+        );
+
+        IBasicIssuanceModule(_tradeData.issuanceModule).issue(_tradeData.setToken, _tradeData.amountSetToken, msg.sender);
+
+        require(inputTokenBalanceBefore.sub(_tradeData.paymentToken.balanceOf(address(this))) <= _tradeData.limitAmount, "FlashMint: OVERSPENT");
+
+        emit FlashMint(msg.sender, _tradeData.setToken, _tradeData.paymentToken, _tradeData.limitAmount, _tradeData.amountSetToken);
+        return inputTokenBalanceBefore.sub(_tradeData.paymentToken.balanceOf(address(this)));
+    }
+
+    /**
+     * Redeem set, redeem fCash components, sell received tokens for output token and transfer proceds to the caller
+     */
+    function _redeemExactSetForToken(
+        TradeData memory _tradeData,
+        DEXAdapter.SwapData[] memory _swapData
+    )
+        internal
+        returns (uint256)
+    {
+
+        uint256 outputTokenBalanceBefore = _tradeData.paymentToken.balanceOf(address(this));
+        if(_tradeData.redeemMaturedPositions) {
+            notionalTradeModule.redeemMaturedPositions(_tradeData.setToken);
+        }
+        _redeemExactSet(_tradeData.setToken, _tradeData.amountSetToken, _tradeData.issuanceModule);
+
+        _redeemWrappedFCashPositions(_tradeData);
+        _sellComponentsForOutputToken(_tradeData, _swapData);
+
+        uint256 outputAmount = _tradeData.paymentToken.balanceOf(address(this)).sub(outputTokenBalanceBefore);
+
+        require(outputAmount >= _tradeData.limitAmount, "FlashMint: UNDERBOUGHT");
+        // Emit event
+        emit FlashRedeem(msg.sender, _tradeData.setToken, _tradeData.paymentToken, _tradeData.amountSetToken, outputAmount);
+        // Return output amount
+        return outputAmount;
+    }
+
+
+    /**
+     * Sells all components (after redemption of fCash positions) for the output token
+     */
+    function _sellComponentsForOutputToken(
+        TradeData memory _tradeData,
+        DEXAdapter.SwapData[] memory _swapData
+    )
+        internal
+    {
+        (address[] memory components, uint256[] memory componentUnits) = _getFilteredComponentsRedemption(
+            _tradeData.setToken,
+            _tradeData.amountSetToken,
+            _tradeData.issuanceModule,
+            _tradeData.isDebtIssuance,
+            _tradeData.slippage
+        );
+
+        require(components.length == _swapData.length, "Components / Swapdata mismatch");
+
+        for (uint256 i = 0; i < components.length; i++) {
+            uint256 maxAmountSell = componentUnits[i];
+            address component = components[i];
+            if(component == address(0)){
+                break;
```

Comment:
No, break is correct here since address zero indicates the end.
The list of filteredComponents (which replaces the fCash components with their underlyings) will typically be shorter than the actual list of components (the fCash addresses).

To avoid having to loop over the components an extra time to determine the length of the filtered components I just create an array of the same length as the original "unfiltered" components, which will be partially empty. As soon as we hit the zero address here we therefore know that we are at the end of the list.

Reply:
See: 
https://github.com/IndexCoop/index-coop-smart-contracts/pull/105/files/6f4228f29646503ab2b7278f71aefc9188fdfb0b#diff-1e7efa5a8ad7c47593c9037f0d3db8656d3caca9d8cadc3a5307c71cf20a5cb6R821

Reply:
I will add a code comment to make this clear though, since I see how this can be confusing.